File: pushy/protocol/proxy.py
# Copyright (c) 2008, 2011 Andrew Wilkins <axwalk@gmail.com>
# 
# Permission is hereby granted, free of charge, to any person
# obtaining a copy of this software and associated documentation
# files (the "Software"), to deal in the Software without
# restriction, including without limitation the rights to use,
# copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following
# conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.

from pushy.protocol.message import message_types, MessageType
import pushy.util
import exceptions, types


class ProxyType(object):
    def __init__(self, code, name):
        self.code = code
        self.name = name
    def __repr__(self):
        return "ProxyType(%d, '%s')" % (self.code, self.name)
    def __str__(self):
        return self.name
    def __int__(self):
        return self.code
    def __hash__(self):
        return self.code
    def __eq__(self, other):
        if type(other) is ProxyType: return other.code == self.code
        if type(other) is int: return other == self.code
        elif type(other) is str: return other == self.name
        return False

    @staticmethod
    def get(obj):
        # First check if it's a class
        class_ = type(obj)
        if hasattr(class_, "pushy_proxy_type"):
            return class_.pushy_proxy_type

        if isinstance(obj, Exception):
            return ProxyType.exception
        if isinstance(obj, dict):
            return ProxyType.dictionary
        if isinstance(obj, list):
            return ProxyType.list
        if isinstance(obj, set):
            return ProxyType.list
        if isinstance(obj, types.ModuleType):
            return ProxyType.module
        if isinstance(obj, types.ClassType):
            return ProxyType.oldstyleclass
        return ProxyType.object

    @staticmethod
    def getargs(proxy_type, obj):
        if type(proxy_type) is int:
            proxy_type = proxy_types[proxy_type]
        if proxy_type is ProxyType.exception:
            module_ = obj.__class__.__module__
            if module_ == "exceptions":
                return obj.__class__.__name__
        return None

    @staticmethod
    def getoperators(obj):
        class_ = type(obj)
        if hasattr(class_, "pushy_operator_mask"):
            return class_.pushy_operator_mask
        else:
            mask = sum((
                (1 << t.code) for t in message_types \
                if t.name.startswith("op__") and hasattr(obj, t.name[2:])
            ))
            return mask


proxy_names = (
  "oldstyleclass",
  "object",
  "exception",
  "dictionary",
  "list",
  "set",
  "module"
)
proxy_types = []
for i,t in enumerate(proxy_names):
    p = ProxyType(i, t)
    proxy_types.append(p)
    setattr(ProxyType, t, p)


###############################################################################


def create_instance(args, conn, on_proxy_init):
    class ClassObjectProxy:
        def __init__(self):
            on_proxy_init(self)
        def __call__(self, *args, **kwargs):
            message_type = pushy.protocol.message.MessageType.op__call__
            return conn.operator(message_type, self, args, kwargs)
        def __getattr__(self, name):
            if name == "__call__":
                raise AttributeError, "__call__"
            return conn.getattr(self, name)
    return ClassObjectProxy


def create_object(args, conn, on_proxy_init):
    class ObjectProxy(object):
        def __init__(self):
            on_proxy_init(self)
            object.__init__(self)
        def __getattribute__(self, name):
            return conn.getattr(self, name)
    return ObjectProxy


def create_exception(args, conn, on_proxy_init):
    BaseException = Exception
    if args is not None:
        BaseException = getattr(exceptions, args)
    class ExceptionProxy(BaseException):
        def __init__(self):
            on_proxy_init(self)
            BaseException.__init__(self)
        def __getattribute__(self, name):
            return conn.getattr(self, name)
    return ExceptionProxy


def create_dictionary(args, conn, on_proxy_init):
    overridden_methods = frozenset(("items", "keys", "update", "values"))
    class DictionaryProxy(dict):
        def __init__(self):
            on_proxy_init(self)
        def keys(self):
            return list(conn.as_tuple(conn.getattr(self, "keys")))
        def items(self):
            return list(conn.as_tuple(conn.getattr(self, "items")))
        def update(self, rhs):
            if type(rhs) is dict:
                conn.getattr(self, "update")(tuple(rhs.items()))
            elif isinstance(rhs, dict):
                conn.getattr(self, "update")(rhs)
            else:
                conn.getattr(self, "update")(tuple(map(tuple, rhs)))
        def values(self):
            return list(conn.as_tuple(conn.getattr(self, "values")))
        def __eq__(self, rhs):
            if self is rhs: return True
            return self.items() == rhs.items()
        def __getattribute__(self, name):
            if name in overridden_methods:
                return object.__getattribute__(self, name)
            return conn.getattr(self, name)
    return DictionaryProxy


def create_list(args, conn, on_proxy_init):
    class ListProxy(list):
        def __init__(self):
            on_proxy_init(self)
            list.__init__(self)
        def __eq__(self, rhs):
            if self is rhs: return True
            as_tuple = conn.eval("tuple")(self)
            rhs_tuple = tuple(rhs)
            return as_tuple == rhs_tuple
        def __getattribute__(self, name):
            return conn.getattr(self, name)
    return ListProxy


def create_set(args, conn, on_proxy_init):
    class SetProxy(set):
        def __init__(self):
            on_proxy_init(self)
            set.__init__(self, args)
        def __getattribute__(self, name):
            return conn.getattr(self, name)
    return SetProxy


def create_module(args, conn, on_proxy_init):
    class ModuleProxy(types.ModuleType):
        def __init__(self):
            on_proxy_init(self)
            types.ModuleType.__init__(self, "")
            self.__importer = None
        def __getattribute__(self, name):
            try:
                # __dict__ on a module is expected to return a dict, not a
                # descendant thereof. Its items will be (in CPython) obtained
                # by ferretting around in the object's internals, rather than
                # going through the usual Python interface.
                result = conn.getattr(self, name)
                if isinstance(result, dict) and name == "__dict__":
                    return dict(result.items())
                return result
            except AttributeError:
                if self.__importer:
                    return self.__importer("%s.%s" % (self.__name__, name))
                else:
                    raise
    return ModuleProxy


class_factories = {
    ProxyType.oldstyleclass: create_instance,
    ProxyType.object:        create_object,
    ProxyType.exception:     create_exception,
    ProxyType.dictionary:    create_dictionary,
    ProxyType.list:          create_list,
    ProxyType.set:           create_set,
    ProxyType.module:        create_module
}

def Proxy(opmask, proxy_type, args, conn, on_proxy_init):
    """
    Create a proxy object, which delegates attribute access and method
    invocation to an object in a remote Python interpreter.
    """

    # Determine the class to use for the proxy type.
    ProxyClass = class_factories[proxy_type](args, conn, on_proxy_init)

    # Store the operator mask and proxy type in the class. This will be used by
    # ProxyType.getoperators() and ProxyType.get()
    #
    # XXX Important! Without this, the performance of tunnelled connections is
    #                unusably slow due to the chatter induced by many calls to
    #                hasattr/getattr.
    ProxyClass.pushy_operator_mask = opmask
    ProxyClass.pushy_proxy_type = proxy_type

    # Creates a lambda with the operator type and connection bound.
    def bound_operator(type_):
        return lambda self, *args, **kwargs: \
            (conn.operator(type_, self, args, kwargs))

    # Operators should be defined if and only if they're not implemented in the
    # local ProxyClass.
    def should_setattr(t):
        if not opmask & (1 << t.code):
            return False
        a = getattr(ProxyClass, t.name[2:], None)
        return a is None or not \
            (type(a) is types.MethodType and a.im_class is ProxyClass)

    # Create proxy operators.
    types_ = (t for t in message_types if should_setattr(t))
    map(lambda t: setattr(ProxyClass, t.name[2:], bound_operator(t)), types_)

    # Add other standard methods.
    setattr(ProxyClass, "__str__", lambda self: conn.getstr(self))
    setattr(ProxyClass, "__repr__", lambda self: conn.getrepr(self))
    setattr(ProxyClass, "__setattr__", lambda *args: conn.setattr(*args))

    # Make sure we support the iteration protocol properly
    #if hasattr(ProxyClass, "__iter__"):
    setattr(ProxyClass, "next", lambda self: conn.getattr(self, "next")())

    return ProxyClass()

File: java/src/pushy/internal/Proxy.java
/*
 * Copyright (c) 2009, 2011 Andrew Wilkins <axwalk@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

package pushy.internal;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Proxy
{
    /**
     * Create an object which provides local access to a remote object.
     */
    public static ProxyObject
    getProxy(Number id, Number operators, Integer type, Object argument,
             Connection connection)
    {
        PushyObjectImpl proxy = new PushyObjectImpl(id, connection);

        // TODO for "object" types, use the operator mask to determine which
        //      interfaces to implement.

        if (Type.list.equals(type))
            return new ProxyList(proxy);

        if (Type.dictionary.equals(type))
            return new ProxyMap(proxy);

        return proxy;
    }

    /**
     * Create an object which 
     */
    public static Type getType(Object object)
    {
        if (object instanceof List)
            return Type.list;

        if (object instanceof Map)
            return Type.dictionary;

        if (object instanceof Set)
            return Type.set;

        if (object instanceof Throwable)
            return Type.exception;

        return Type.object;
    }

    /**
     * Get a bitmask for the operators supported by an object.
     */
    public static Number getOperators(Object object)
    {
        java.math.BigInteger mask = new java.math.BigInteger("0");
        for (Iterator iter = Message.Type.getTypes().iterator();
             iter.hasNext();)
        {
            Message.Type type = (Message.Type)iter.next();
            if (type.getName().startsWith("op__"))
            {
                if (hasOperator(object, type))
                    mask = mask.setBit(type.getCode());
            }
        }
        return mask;
    }

    /**
     * Check if the object supports the specified operator.
     */
    public static boolean hasOperator(Object object, Message.Type type)
    {
        // __cmp__ maps to Comparable.
        if (type.equals(Message.Type.op__cmp__) ||
            type.equals(Message.Type.op__rcmp__))
        {
            return object instanceof Comparable;
        }

        // Everything supports __hash__.
        if (type.equals(Message.Type.op__hash__))
            return true;

        // Collections support __len__.
        if (type.equals(Message.Type.op__len__))
            return object instanceof Collection;

        // Lists and maps support __getitem__, __setitem__ and __contains__.
        if (type.equals(Message.Type.op__getitem__) ||
            type.equals(Message.Type.op__setitem__) ||
            type.equals(Message.Type.op__contains__))
        {
            return object instanceof List || object instanceof Map;
        }

        // All collections support __delitem__.
        if (type.equals(Message.Type.op__delitem__))
            return object instanceof Collection;

        // Iterables support __iter__.
        if (type.equals(Message.Type.op__iter__))
            return object instanceof Iterable;

        // Classes (constructors) and BoundMethod support __call__.
        if (type.equals(Message.Type.op__call__))
            return object instanceof Class || object instanceof Callable;

        return false;
    }

    /**
     * Get argument for the proxy object, to pass to the remote side.
     */
    public static Object getArgument(Object object, Type type)
    {
/*
        if (type.equals(Type.dictionary))
        {
            Map map = (Map)object;
            Object[] items = new Object[map.size()];
            int i = 0;
            for (Iterator iter = map.entrySet().iterator(); iter.hasNext();)
            {
                Map.Entry entry = (Map.Entry)iter.next();
                items[i++] = new Object[]{entry.getKey(), entry.getValue()};
            }
            return items;
        }
        else if (type.equals(Type.list) || type.equals(Type.set))
        {
            return ((Collection)object).toArray(new Object[]{});
        }
*/
        return null;
    }

    /**
     * A class for describing a proxied object type.
     */
    public static class Type
    {
        private int code;
        private String name;

        public Type(int code, String name)
        {
            this.code = code;
            this.name = name;
        }

        public int getCode()
        {
            return code;
        }

        public String getName()
        {
            return name;
        }

        public String toString()
        {
            return "ProxyType(" + code + ", '" + name + "')";
        }

        public int hashCode()
        {
            return code;
        }

        public boolean equals(Object rhs)
        {
            if (rhs instanceof Type)
                return code == ((Type)rhs).code;
            else if (rhs instanceof Integer)
                return code == ((Integer)rhs).intValue();
            else if (rhs instanceof String)
                return name.equals(rhs);
            return super.equals(rhs);
        }

        // Provide a means of getting a message type by its code.
        private static List types = new ArrayList();
        public static Type getType(int code)
        {
            return (Type)types.get(code);
        }

        // Method for defining a type, given a name. The type's code
        // will be the next index into the 'types' list.
        private static Type createType(String name)
        {
            Type type = new Type(types.size(), name);
            types.add(type);
            return type;
        }

        // Define proxy types: must be in the same order as in the Python
        // code.
        public static final Type oldstyleclass = createType("oldstyleclass");
        public static final Type object        = createType("object");
        public static final Type exception     = createType("exception");
        public static final Type dictionary    = createType("dictionary");
        public static final Type list          = createType("list");
        public static final Type set           = createType("set");
        public static final Type module        = createType("module");
    }
}

